### PR TITLE
Add fusion regression

### PR DESCRIPTION
**Pull request by Stefan824**
This PR should address issue #2595 if goes well.

- Added a new script `run_fusion_regression.py` to handle fusion regression tasks. It behaves similarly to `run_regression.py`.
- Included two test yaml files.
- Tests could be performed following ```anserini/docs/fuse-regressions/test-setup.md```.
- Next steps in mind are 
  - Implementing an auto-script to generate more yaml files for fusion based on current regression yamls (working on it)
  - Doing experiments to find out expected eval results for fusion on different runs with different fusion methods and parameters

### CODE REVIEW

File: src/main/resources/fuse_regression/test_1.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+topic_reader: TsvString
+topics:
+  - name: "BEIR (v1.0.0): Robust04"
+    id: test
+    path: topics.beir-v1.0.0-robust04.test.tsv.gz
+    qrel: qrels.beir-v1.0.0-robust04.test.txt
+
+# Fusion Regression Test Configuration
+runs:
```
Comment (lintool) [inline src/main/resources/fuse_regression/test_1.yaml:36]: Why don't we assume that the runs are generated by a regression in `https://github.com/castorini/anserini/tree/master/src/main/resources/regression`? E.g., `beir-v1.0.0-robust04.flat.yaml`

So we keep the yaml file and the run name (i.e., just take the run file name it generates).

File: src/main/resources/regression/beir-v1.0.0-robust04.flat.yaml [hunk 2]
```diff
@@ -53,3 +53,4 @@ models:
         - 0.3746
       R@1000:
         - 0.6345
+
```
Comment (lintool) [inline src/main/resources/regression/beir-v1.0.0-robust04.flat.yaml:56]: revert this change, i.e., don't unnecessarily change files

File: src/main/resources/fuse_regression/test_0.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
+---
+corpus: beir-v1.0.0-robust04.flat
```
Comment (lintool) [inline src/main/resources/fuse_regression/test_0.yaml:2]: rename file to be more descriptive? e.g., `beir-v1.0.0-robust04.flat`?

File: pom.xml [hunk 4]
```diff
@@ -704,5 +704,17 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
```
Comment (lintool) [inline pom.xml:707]: do we really need these? I recall this causing issues previously...

Comment (lintool) [reply pom.xml:707]: https://github.com/castorini/anserini/pull/2597

**Review by lintool — CHANGES_REQUESTED** [review]
Initial comments.

File: src/main/resources/fuse_regression/test_1.yaml (hunk 1 above)
Comment (Stefan824) [reply src/main/resources/fuse_regression/test_1.yaml:36]: I established a test_2.yaml, do you mind taking a look to see if that is something you'd like? 
## Perform two regression runs for test2
One could generate the runs necessary for test 2 following 
- https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-beir-v1.0.0-robust04.bge-base-en-v1.5.flat-int8.cached.md
- https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-beir-v1.0.0-robust04.bge-base-en-v1.5.flat.cached.md

Or do you want the fusion yaml file to contain information about the yaml files (the ones that produce runs for the fusion test) it depends on and perform those runs automatically when the fusion test is performed?

File: pom.xml (hunk 4 above)
Comment (Stefan824) [reply pom.xml]: Did this modification on a different branch. Forgot to merge. Sorry and it's now resolved.

File: src/main/resources/regression/beir-v1.0.0-robust04.flat.yaml (hunk 2 above)
Comment (Stefan824) [reply src/main/resources/regression/beir-v1.0.0-robust04.flat.yaml]: Done

File: src/main/resources/fuse_regression/test_1.yaml (hunk 1 above)
Comment (Stefan824) [reply src/main/resources/fuse_regression/test_1.yaml]: Its name is now changed into `src/main/resources/fuse_regression/fusion-regression-bge-flat-int8-robust04-2.yaml`

File: src/main/resources/regression/beir-v1.0.0-robust04.flat.yaml [hunk 5]
```diff
@@ -52,4 +52,4 @@ models:
       R@100:
         - 0.3746
       R@1000:
-        - 0.6345
+        - 0.6345
```
Comment (lintool) [inline src/main/resources/regression/beir-v1.0.0-robust04.flat.yaml:55]: revert unnecessary changes please

File: src/main/resources/fuse_regression/fusion-regression-bge-flat-robust04.yaml-2.yaml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+topic_reader: TsvString
+topics:
+  - name: "BEIR (v1.0.0): Robust04"
+    id: test
+    path: topics.beir-v1.0.0-robust04.test.tsv.gz
+    qrel: qrels.beir-v1.0.0-robust04.test.txt
+
+# Fusion Regression Test Configuration
+runs:
+  - runs/runs.beir/run.inverted.beir-v1.0.0-robust04.flat.test.bm25
```
Comment (lintool) [inline src/main/resources/fuse_regression/fusion-regression-bge-flat-robust04.yaml-2.yaml:37]: record the regression yaml this run comes from?

File: docs/fuse-regressions/test-setup.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+To automatically download the required files, you can use the following shell script. The script will download and extract the files in the `runs/runs.beir` folder with the correct filenames.
+
+```bash
+#!/bin/bash
+
+# Create the target directory if it doesn't exist
+mkdir -p runs/runs.beir
+
+# Download the run files from Google Drive using their file IDs
+wget --no-check-certificate 'https://drive.google.com/uc?export=download&id=1XVlVCDYQe3YjRzxplaeGbmW_0EFQCgm8' -O runs/runs.beir/run.inverted.beir-v1.0.0-robust04.multifield.test.bm25
```
Comment (lintool) [inline docs/fuse-regressions/test-setup.md:21]: depend on the runs generated by the regression yaml config?

File: src/main/resources/fuse_regression/fusion-regression-bge-flat-int8-robust04-2.yaml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  - runs/run.flat.beir-v1.0.0-robust04.bge-base-en-v1.5.test.bge-flat-cached
+methods:
+  - name: rrf
+    k: 1000
+    depth: 1000
+    rrf_k: 60
+    output: runs/fuse/run.flat-int8.beir-v1.0.0-robust04.bge-base-en-v1.5.test.bge-flat-int8-cached.bge-flat-cached.fusion.rrf
+    results:
+      nDCG@10:
+        - 0.3
```
Comment (lintool) [inline src/main/resources/fuse_regression/fusion-regression-bge-flat-int8-robust04-2.yaml:47]: record to 4 digits?

**Review by lintool — CHANGES_REQUESTED** [review]

**Review by lintool — CHANGES_REQUESTED** [review]
Why don't we start with fusion robust04 BM25 and dense:

BM25: https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-beir-v1.0.0-robust04.flat.md

BGE: https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-beir-v1.0.0-robust04.bge-base-en-v1.5.flat.onnx.md

Accessible here: https://github.com/castorini/anserini?tab=readme-ov-file#beir-v100-regressions

**Comment by Stefan824** [comment]
> Why don't we start with fusion robust04 BM25 and dense:
> 
> BM25: https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-beir-v1.0.0-robust04.flat.md
> 
> BGE: https://github.com/castorini/anserini/blob/master/docs/regressions/regressions-beir-v1.0.0-robust04.bge-base-en-v1.5.flat.onnx.md
> 
> Accessible here: https://github.com/castorini/anserini?tab=readme-ov-file#beir-v100-regressions

Added corresponding yaml. Do you like the structure of the yaml and the namings? If so, I could start on generating more.

File: src/main/resources/fuse_regression/fusion-regression-bge-flat-int8-robust04-2.yaml [hunk 9]
```diff
@@ -0,0 +1,79 @@
+---
```
Comment (lintool) [inline src/main/resources/fuse_regression/fusion-regression-bge-flat-int8-robust04-2.yaml:1]: let's remove this and keep only the above?

File: src/main/resources/fuse_regression/beir-v1.0.0-robust04.flat.bm25.fuse.bge-base-en-v1.5.bge-flat-onnx.yaml [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+runs:
+  - runs/run.beir-v1.0.0-robust04.flat.bm25.topics.beir-v1.0.0-robust04.test.txt
+  - runs/run.beir-v1.0.0-robust04.bge-base-en-v1.5.bge-flat-onnx.topics.beir-v1.0.0-robust04.test.txt
+
+methods:
+  - name: rrf
+    k: 1000
+    depth: 1000
+    rrf_k: 60
+    output: runs/fuse/run.beir-v1.0.0-robust04.flat.bm25.fuse-rrf.bge-base-en-v1.5.bge-flat-onnx.topics.beir-v1.0.0-robust04.test.txt
```
Comment (lintool) [inline src/main/resources/fuse_regression/beir-v1.0.0-robust04.flat.bm25.fuse.bge-base-en-v1.5.bge-flat-onnx.yaml:45]: instead of a subdir, how about `runs/run.fuse.XXXX`?

Comment (lintool) [reply src/main/resources/fuse_regression/beir-v1.0.0-robust04.flat.bm25.fuse.bge-base-en-v1.5.bge-flat-onnx.yaml:45]: How about `runs/runs.fuse.rrf.XXX` and `runs/runs.fuse.avg.XXX`

**Review by lintool — CHANGES_REQUESTED** [review]

**Review by lintool — APPROVED** [review]

**Comment by lintool** [comment]
hey @Stefan824 this is good for merging, but I pinged you on Slack for improvements. Please work on that next. Thanks!
